File: E_Kefa_and_Park.py
import sys
sys.setrecursionlimit(2 * (10 ** 5))
n, m = map(int, input().split())
cats = list(map(int, input().split()))
tree = [[] for _ in range(n)]



for i in range(n - 1):
    x, y = map(int, input().split())
    tree[x - 1].append(y - 1)
    tree[y - 1].append(x - 1)

def recur(node, path, parent):
    catsCount = 0
    if cats[node] == 0:
        catsCount = 0
    else:
        catsCount = path + cats[node]
    if catsCount > m:
        return 0
    leaf = True
    valid = 0
    for j in tree[node]:
        if j == parent:
            continue
        leaf = False
        valid += recur(j, catsCount, node)
    if leaf:
        return 1
    return valid

print(recur(0, 0, -1))

File: E_Kefa_and_Park.cpp
#include<map>
#include<iostream>
#include<vector>
using namespace std;

int m;
vector<int> cats;
vector<vector<int> > tree;

int recur(int node, int path, int parent) {
    int catCount = 0, valid = 0;
    bool leaf = true;
    if(cats[node] == 0) {
        catCount = 0;
    } else {
        catCount = path + cats[node];
    }
    if(catCount > m) {
        return 0;
    }
    for(int i = 0; i < tree[node].size(); i++) {
        if(tree[node][i] == parent) {
            continue;
        }
        leaf = false;
        valid += recur(tree[node][i], catCount, node);
    }
    if(leaf) {
        return 1;
    }
    return valid;
}

int main() {
    int n,temp,x,y;
    cin >> n >> m;
    for(int i = 0; i < n; i++) {
        cin >> temp;
        cats.push_back(temp);
        tree.push_back(vector<int>());
    }
    for(int i = 0; i < n - 1; i++) {
        cin >> x >> y;
        tree[x - 1].push_back(y - 1);
        tree[y - 1].push_back(x - 1);
    }
    int result = recur(0, 0, -1);
    cout << result << endl;
    return 0;
}